### PR TITLE
Adding "Azure_CloudService_SI_Disable_RemoteDesktop_Access" remediati…

### PR DESCRIPTION
**Pull request by RohitYadav-msft**
…on script

### CODE REVIEW

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    This script is used to disable remote Desktop (RDP) access on cloud service(s).
+
+ControlId: 
+    Azure_CloudService_SI_Disable_RemoteDesktop_Access
+
+DisplayName:
+    Remote Desktop (RDP) access must be disabled on cloud service roles.    
+
+# Pre-requisites:
+    You will need co-administrator role on subscription.
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:13]: Please confirm if just a `Contributor` role would suffice.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+#>
+
+function Pre_requisites
+{
+    <#
+    .SYNOPSIS
+    This command would check pre requisites modules.
+    .DESCRIPTION
+    This command would check pre requisites modules to perform remediation.
+	#>
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:53]: Nit: Please check the indentation.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+
+    Write-Host "Required modules are: Az.Account, Az.Resources, Azure" -ForegroundColor Cyan
+    Write-Host "Checking for required modules..."
+    $availableModules = $(Get-Module -ListAvailable Az.Resources, Az.Accounts, Azure)
+    
+    # Checking if 'Az.Accounts' module is available or not.
+    if($availableModules.Name -notcontains 'Az.Accounts')
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:60]: Can lines 59 - 90 replaced with a loop?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    $totalCloudService = ($resources | Measure-Object).Count
+    if($totalCloudService -eq 0)
+    {
+        Write-Host "No cloud service(s) found." -ForegroundColor $([Constants]::MessageType.Error);
+        Write-Host $([Constants]::DoubleDashLine)
+        break;
+    }
+
+    Write-Host "Total cloud service(s): [$($totalCloudService)]"
+    $folderPath = [Environment]::GetFolderPath("MyDocuments")
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:218]: Do we change the path to `%temp% or the current directory?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        $cloudServiceWithDisabledRDPAccess = @()
+        $skippedCloudServiceFromRemediation = @()
+
+        # Checking cloud service(s) with enabled RDP acccess
+        $resources |  ForEach-Object {
+            $resource = $_
+            $rdpExtensions = $()
+            $cloudServiceExtension = Get-AzureServiceRemoteDesktopExtension -ServiceName $_.Name -ErrorAction SilentlyContinue
+            
+            $rdpExtensions += $cloudServiceExtension | Where-Object { ($null -ne $_) -and ($_.Extension -like "*RDP*") }
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:238]: Do we need to check only "enabled" extensions?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                $item =  New-Object psobject -Property @{  
+                    CloudServiceName = $_.Name                
+                    ResourceGroupName = $_.ResourceGroupName
+                    RDPExtensionDetails = $rdpExtensions
+                }
+                    
+                $cloudServiceWithEnabledRDPAccess += $item        
+            }
+            else {
+                $cloudServiceWithDisabledRDPAccess += $_ | Select-Object @{Expression={($resource.ResourceGroupName)};Label="ResourceGroupName"},@{Expression={$resource.Name};Label="CloudServiceName"}
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:251]: There can be two cases:
- Extension is not present.
- Extension is present, but, disabled.
Do we consider both of them for the "disabled" use case?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+}
+
+class Constants
+{
+    static [Hashtable] $MessageType = @{
+        Error = [System.ConsoleColor]::Red
+        Warning = [System.ConsoleColor]::Yellow
+        Info = [System.ConsoleColor]::Cyan
+        Update = [System.ConsoleColor]::Green
+	    Default = [System.ConsoleColor]::White
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:350]: Nit: Please fix the indentation.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+                {
+                    return;
+                }
+            }
+            
+            Write-Host "`t"
+            Write-Host "Disabling RDP access on [$($totalCloudServiceWithEnableRDPAccess)] cloud service(s)..."
+            try 
+            {
+                $cloudServiceWithEnabledRDPAccess | ForEach-Object {
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:292]: Please document that the script does not currently remediate Cloud Services that had RDP enabled during the deployment, in which case, no extensions would be present.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 1 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: We can execute remediation script using contributor permission, have added changes for this

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 2 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:53]: Indentation looks good in VS, 
will resolve this comments at the end from portal

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 3 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: Done

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 4 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: Done

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 5 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: Yes

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 8 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:285]: Added

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 7 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:349]: Will do at the end from portal

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 6 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:244]: Covered this  scenario.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+                    $_.RDPExtensionDetails | ForEach-Object {
+                        $output = Remove-AzureServiceRemoteDesktopExtension -ServiceName $serviceName -Slot $_.Slot -Role $_.Role -ErrorAction SilentlyContinue
+                        if($null -eq $output -and $output.OperationStatus -ine "Succeeded")
+                        {
+                            $item =  New-Object psobject -Property @{  
+                                CloudServiceName = $serviceName                
+                                ResourceGroupName = $rgName
+                            }
+    
+                            $skippedCloudServiceFromRemediation += $item
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:299]: Are you breaking on first failure?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+                            break;
+                        }
+                    }
+
+                    if(($skippedCloudServiceFromRemediation | Measure-Object).Count -eq 0)
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:304]: Are lines 304-307 valid?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+                    {
+                        $_ | Select-Object @{Expression={($_.ResourceGroupName)};Label="ResourceGroupName"},@{Expression={$_.CloudServiceName};Label="CloudServiceName"} | Sort-Object | Format-Table | Out-String
+                    }
+                }
+            }
+            catch
+            {
+                $item =  New-Object psobject -Property @{  
+                    CloudServiceName = $_.CloudServiceName                
+                    ResourceGroupName = $_.ResourceGroupName
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:314]: Do you want to include the extension details here?

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]
Please include another reviewer.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 9 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:299]: We have used 'For' loop to remove extensions from cloud service(s), so if we failed to remove any extension(due to permission issue or cmd. failure) then we are breaking flow of removing other extension of that cloud service and simply storing service and RG name in an array(containing list of cloud services skipped from remediation).

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 10 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:304]: Yes, here we are printing table, containing service and RG name on which we have successfully performed remediation.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 11 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:314]: Here we are showing service and RG name of cloud services for which we are unable to do remediation(due to permission issue or any command failure).
Therefore showing extension details is not required

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+                }
+            }
+            
+            Write-Host "`t"
+            Write-Host "Disabling RDP access on [$($totalCloudServiceWithEnableRDPAccess)] cloud service(s)..."
+            try 
+            {
+                $cloudServiceWithEnabledRDPAccess | ForEach-Object {
+                    # Disabling RDP access at service level
+                    $output = $_.RDPExtensionDetails | Remove-AzureServiceRemoteDesktopExtension -ServiceName $_.CloudServiceName -ErrorAction SilentlyContinue
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: Do you want to provide the `UninstallConfiguration` flag to this command?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+        Write-Host "Connecting to AzAccount..."
+        Connect-AzAccount -ErrorAction Stop
+        Write-Host "Connected to AzAccount" -ForegroundColor $([Constants]::MessageType.Update)
+    }
+
+    # Setting context for current subscription.
+    $currentSub = Set-AzContext -SubscriptionId $SubscriptionId -Force -ErrorAction Stop
+    Select-AzureSubscription -SubscriptionId $($SubscriptionId) -ErrorAction Stop
+    
+    Write-Host "Note: `n Cloud services on which RDP was enabled, during the deployment will not be remediated." -ForegroundColor Yellow
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:131]: Please include a note that says - These need to be remediated via Azure Portal.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+    Write-Host "------------------------------------------------------"
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:132]: Is rollback supported? If not yet, and if rollback is not possible, please include a prompt that asks for the user's input before proceeding.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 14 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:132]: Added

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 13 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: Added

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 12 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: Not considering for now, we can take it as enhancement.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+            $rdpExtensions = @()
+            
+            # Fetching cloud service(s) slot i.e. Staging, Production 
+            $cloudServiceSlots = Get-AzResource -ResourceGroupName $resource.ResourceGroupName -Name $resource.Name -ResourceType "$resourceType/slots" -ApiVersion "2016-04-01"
+            $cloudServiceSlotsName = $cloudServiceSlots.Name | sort -Unique
+            
+            # Checking RDP extension on each slots
+            $cloudServiceSlotsName | ForEach-Object {
+                $slot = $_
+                $cloudServiceExtension = Get-AzureServiceRemoteDesktopExtension -ServiceName $resource.Name -Slot $slot -ErrorAction SilentlyContinue
```
Comment (TarunKrShukla) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:228]: Does this return RDP extension for all roles in a slot?

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+        Write-Host "Connecting to AzAccount..."
+        Connect-AzAccount -ErrorAction Stop
+        Write-Host "Connected to AzAccount" -ForegroundColor $([Constants]::MessageType.Update)
+    }
+
+    # Setting context for current subscription.
+    $currentSub = Set-AzContext -SubscriptionId $SubscriptionId -Force -ErrorAction Stop
+    Select-AzureSubscription -SubscriptionId $($SubscriptionId) -ErrorAction Stop
+    
+    Write-Host "Note: `n Cloud services on which RDP was enabled, during the deployment will not be remediated(For now these need to be remediated via azure portal)." -ForegroundColor $([Constants]::MessageType.Warning)
```
Comment (TarunKrShukla) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:131]: We should output list of such cloud services.

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 15 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:228]: Yes

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 (hunk 16 above)
Comment (RohitYadav-msft) [reply Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1]: Girish will be working on this to output list of service on which RDP enabled while doing deployment

**Review by gvaradarajan-msft — APPROVED** [review]
Please get a second review.

**Review by TarunKrShukla — APPROVED** [review]

File: Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1 [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+        Write-Host "Error occured while performing remediation. ErrorMessage [$($_)]" -ForegroundColor $([Constants]::MessageType.Error)
+        break
+    }
+
+    if(($skippedCloudServiceFromRemediation | Measure-Object).Count -eq 0)
+    {
+        Write-Host "Successfully disabled RDP access from cloud service(s)." -ForegroundColor $([Constants]::MessageType.Update)
+    }
+    else 
+    {
```
Comment (TarunKrShukla) [inline Scripts/RemediationScripts/Remediate-RemoteDesktopAccess.ps1:338]: In this case for some of the cloud services we might have be able to remove RDP extensions we should print those as well.

**Review by TarunKrShukla — CHANGES_REQUESTED** [review]
Please work on comments.

**Review by TarunKrShukla — APPROVED** [review]

**Pull request merged**